File: GTAClone-iOS/GTAClone/entities/Entity3D.swift
import SceneKit

class Entity3D {
    var x: Float; var y: Float = 0; var z: Float
    var rotY: Float = 0
    var health: Float = 100
    let node: SCNNode

    init(x: Float, z: Float) {
        self.x = x; self.z = z
        node = SCNNode()
    }

    func syncNode() {
        node.position   = SCNVector3(x, y, z)
        node.eulerAngles.y = rotY * .pi / 180
    }

    func distanceTo(_ o: Entity3D) -> Float { hypot(x - o.x, z - o.z) }
    func distanceTo(x ox: Float, z oz: Float) -> Float { hypot(x - ox, z - oz) }
}

File: app/src/main/kotlin/com/gtaclone/entities/Entity3D.kt
package com.gtaclone.entities

abstract class Entity3D(var x: Float, var y: Float, var z: Float) {
    var rotY = 0f
    var health = 100f
    var vx = 0f; var vz = 0f

    abstract val w: Float; abstract val h: Float; abstract val d: Float
    abstract fun color(): FloatArray

    fun distanceTo(o: Entity3D) = kotlin.math.hypot(x - o.x, z - o.z)
    fun distanceTo(ox: Float, oz: Float) = kotlin.math.hypot(x - ox, z - oz)
}
